src/python/pants/engine/process.py

```diff
@@ -513,14 +513,27 @@ async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
         bash = await Get(BashBinary, BashBinaryRequest())
         shebang = f"#!{bash.path}"
 
-    # Note: the backslash after the """ marker ensures that the shebang is at the start of the
-    # script file. Many OSs will not see the shebang if there is intervening whitespace.
+    # Some subtle notes with this script:
+    #
+    #  - The backslash after the `"""` ensures that the shebang is at the the backslash after the
+    #       """ marker ensures that the shebang is at the start of the script file. Many OSs will
+    #       not see the shebang if there is intervening whitespace.
+    #  - We run the script with `ProcessResult` instead of `FallibleProcessResult` so that we
+    #      can catch bugs in the script itself, given an earlier silent failure.
+    #  - We do not use `set -e` like normal because it causes the line
+    #      `command which -a <bin> || true` to fail the script when using Bash 3, which macOS
```

Comment:
Thanks for confirming this. The 3 to 5 gap is huge enough the explanation seems correct.